### PR TITLE
Mg/md

### PR DESCRIPTION
**Pull request by mgastegger**
MD code for new SchNetPack.

### CODE REVIEW

File: src/schnetpack/md/md_configs/dynamics/barostat/nhc_aniso.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+_target_: nhc_barostat_aniso
```
Comment (ktschuett) [inline src/schnetpack/md/md_configs/dynamics/barostat/nhc_aniso.yaml:1]: use exlicit clases here

File: src/schnetpack/md/md_configs/calculator/schnetpack.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+_target_: schnetpack
```
[the remark this answers is not in the document]
Comment (ktschuett) [reply src/schnetpack/md/md_configs/calculator/schnetpack.yaml:1]: not a good alias

**Review by ktschuett — CHANGES_REQUESTED** [review]

File: src/schnetpack/md/simulation_hooks/barostats.py [hunk 3]
```diff
@@ -0,0 +1,904 @@
...
+        # Compute sinh(x)/x term
+        b_coeff = a_coeff * self.sinhdx.f(scaled_velocity)
+
+        # Update the momenta (using half timestep)
+        system.momenta = (
+            system.momenta * system.expand_atoms(a_coeff ** 2)
+            + system.forces * system.expand_atoms(b_coeff) * self.time_step * 0.5
+        )
+
+    # def compute_conserved(self, system):
```
[the remark this answers is not in the document]
Comment (mgastegger) [reply src/schnetpack/md/simulation_hooks/barostats.py:691]: I would prefer to leave this in for future debugging. This computes the conserved quantity of the thermostat to check whether it is implemented correctly. It is not necessary for production runs.

File: src/schnetpack/md/utils/base.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+import logging
+
+log = logging.getLogger(__name__)
+
+
+def set_random_seed(seed):
```
[the remark this answers is not in the document]
Comment (mgastegger) [reply src/schnetpack/md/utils/base.py:6]: Deleted function and module. The MD was already using `pl.seed_everything`.

File: src/schnetpack/md/simulator.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+        # Check, if computational graph should be built
+        if self.gradients_required:
+            grad_context = torch.no_grad()
+        else:
+            grad_context = nullcontext()
+
+        with grad_context:
+            # Perform initial computation of forces
+            self.calculator.calculate(self.system)
```
Comment (ktschuett) [inline src/schnetpack/md/simulator.py:117]: `self.system` could turn out to be a problem...

File: src/schnetpack/md/md_configs/calculator/schnetpack.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgastegger) [reply src/schnetpack/md/md_configs/calculator/schnetpack.yaml]: Completely removed alias

File: src/schnetpack/md/md_configs/calculator/schnetpack_ensemble.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+_target_: schnetpack_ensemble
```
[the remark this answers is not in the document]
Comment (mgastegger) [reply src/schnetpack/md/md_configs/calculator/schnetpack_ensemble.yaml]: Removed alias

File: src/schnetpack/md/md_configs/dynamics/barostat/nhc_aniso.yaml (hunk 1 above)
Comment (mgastegger) [reply src/schnetpack/md/md_configs/dynamics/barostat/nhc_aniso.yaml]: removed alias

File: src/schnetpack/md/simulator.py (hunk 5 above)
Comment (mgastegger) [reply src/schnetpack/md/simulator.py:117]: Keep as is for now

**Comment by mgastegger** [comment]
- Completely removed alias functions
- Restructured configs
- Added pytorch 1.9.0 lazy init functionality for registering buffers properly
- Fixed restarts
- Added logging of pressure, potential energy and stress to HDF5 and added read to loader

File: src/schnetpack/md/calculators/schnetpack_calculator.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+from schnetpack.md.calculators.base_calculator import MDCalculator, MDCalculatorError
+from schnetpack.md.calculators.ensemble_calculator import EnsembleCalculator
+
+log = logging.getLogger(__name__)
+
+__all__ = ["SchnetPackCalculator", "SchnetPackEnsembleCalculator"]
+
+
+class SchnetPackCalculator(MDCalculator):
```
Comment (ktschuett) [inline src/schnetpack/md/calculators/schnetpack_calculator.py:22]: SchNetPack with capital N :)

File: src/schnetpack/atomistic/model.py [hunk 8]
```diff
@@ -95,10 +95,13 @@ def __init__(
...
 
-        self.required_derivatives = set()
+        self.required_derivatives = []
         for m in self.output_modules:
             if hasattr(m, "required_derivatives"):
-                self.required_derivatives.update(m.required_derivatives)
+                for p in m.required_derivatives:
+                    if p not in self.required_derivatives:
+                        self.required_derivatives.append(p)
+
```
Comment (ktschuett) [inline src/schnetpack/atomistic/model.py:104]: couldn't you keep it a set and then do: `self.required_derivatives=list(required_derivatives)`?

File: src/schnetpack/md/calculators/schnetpack_calculator.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        Returns:
+            dict(torch.Tensor): Schnetpack inputs in dictionary format.
+        """
+        inputs = self._get_system_molecules(system)
+        neighbors = self.neighbor_list.get_neighbors(inputs)
+        inputs.update(neighbors)
+        return inputs
+
+
+class SchnetPackEnsembleCalculator(EnsembleCalculator, SchnetPackCalculator):
```
Comment (ktschuett) [inline src/schnetpack/md/calculators/schnetpack_calculator.py:176]: N!

**Review by ktschuett — APPROVED** [review]

File: src/schnetpack/atomistic/model.py (hunk 8 above)
Comment (mgastegger) [reply src/schnetpack/atomistic/model.py:104]: This was a leftover from the torch script trial. Reverted it to original set version

**Pull request merged**